File: ukidelly/PGS/240604/배열_조각하기.java
import java.util.Arrays;

/**
 * 결국 푸는 방법이 생각 안나서 정답을 봤습니다. 그 중 가장 인상적인 정답이더군요.
 * 배열을 실시간으로 자르는 것이 아닌, 시작과 끝 인덱스를 저장하고, 마지막에 Arrays.copyOfRange()로 잘라서 반환하는 방법이었습니다.
 */
class Solution {
  public int[] solution(int[] arr, int[] query) {
    int start = 0;
    int end = arr.length - 1;
    for (int i = 0; i < query.length; i++) {
      if (i % 2 == 0) {
        // i가 짝수일 때, 뒤를 버리는 것이니까, start에서부터 query[i] - 1 인덱스를 제외하고 버림
        end = start + query[i] - 1;
      } else {
        // i가 홀수 이면, start의 위치를 query[i]로 이동
        start += query[i];
      }
    }

    return Arrays.copyOfRange(arr, start, end + 2);
  }
}

File: ukidelly/PGS/240604/배열_조각하기.kt
class Solution {

  /**
   * 제가 푼 방식
   */
  fun solution(arr: IntArray, query: IntArray): List<Int> {

    var list = arr.toList()

    query.forEach {
      list = if (it % 2 == 0) {
        list.slice(0..<it)
      } else {
        list.slice(it..<list.lastIndex)
      }
    }

    return list
  }

  // Java에서 푼 정답을 보고 Kotlin으로 적용한 방식
  fun solution2(arr: IntArray, query: IntArray): List<Int> {
    val arrList = arr.toList()

    var (startIndex, endIndex) = 0 to arrList.lastIndex

    for (q in query) {
      // q가 짝수일때, endIndex를 startIndex에서 q의 이전 인덱스로 지정 ( q를 포함해서 뒤 요소들을 버려야하므로 )
      if (q.isEven()) endIndex = startIndex + (q - 1)
      // q가 홀수 일때, startIndex를 q만큰 앞으로 이동 ( q를 포함해서 앞 요소들을 버려야하기 떄문 )
      else startIndex += q
    }
    
    return arrList.slice(startIndex..endIndex)
  }
}

// 짝수인지 홀수인지 판별하는 확장함수
fun Int.isEven() = this % 2 == 0